pdp/target.go

```diff
@@ -196,7 +196,11 @@ var TargetCompatibleExpressions = map[string]map[Type]map[Type]twoArgumentsFunct
 		TypeInteger: {
 			TypeInteger: makeFunctionIntegerEqual},
 		TypeFloat: {
-			TypeFloat: makeFunctionFloatEqual}},
+			TypeFloat: makeFunctionFloatEqual},
```

Comment:
This won't work. You registered function ListOfStringsEqual which takes two lists of strings as arguments as a function which suitable for TypeListOfStrings and TypeInteger argument pair. Similarly, makeFunctionSetOfStringsEqual won't work with TypeSetOfStrings and TypeFloat pair. Please remove all the new "equal" functions from here or put correct argument types:
```go
var TargetCompatibleExpressions = map[string]map[Type]map[Type]twoArgumentsFunctionType{
	"equal": {
		...
		TypeListOfStrings: {
 			TypeListOfStrings: makeFunctionListOfStringsEqual},
		TypeSetOfStrings: {
			TypeSetOfStrings: makeFunctionSetOfStringsEqual}},
	},
	...
}
```

Reply:
Also you can make following change in `github.com/infobloxopen/themis/pdp/ast/yast/parser_test.go` (parset_test.go contains special policy `allFeaturePolicies` to test all possible cases):
```
$ git diff
diff --git a/pdp/ast/yast/parser_test.go b/pdp/ast/yast/parser_test.go
index 077bee5d..3f30ec46 100644
--- a/pdp/ast/yast/parser_test.go
+++ b/pdp/ast/yast/parser_test.go
@@ -204,6 +204,16 @@ policies:
       - val:
          type: string
          content: string
+    - equal:
+      - attr: lsAttr
+      - val:
+         type: list of strings
+         content: [first, second, third]
+    - equal:
+      - attr: ssAttr
+      - val:
+         type: set of strings
+         content: [first, second, third]
     - all:
       - contains:
         - val:
```
With the change I get following in the directory:
```
$ go test -run TestUnmarshal
--- FAIL: TestUnmarshal (0.00s)
    parser_test.go:935: Expected no error but got *yast.matchFunctionCastError (#24 (hidden policy set>target>2>any>3): No function equal for arguments List of Strings and List of Strings)
FAIL
exit status 1
FAIL	github.com/infobloxopen/themis/pdp/ast/yast	0.037s
```
This is exactly because I don't have "equal" function to those arguments. With your changes it should work.

Similar change you can make in `github.com/infobloxopen/themis/pdp/ast/jast/parser_test.go` and check if JAST parser understands you correctly:
```
$ git diff
diff --git a/pdp/ast/jast/parser_test.go b/pdp/ast/jast/parser_test.go
index cca078a0..1fed525f 100644
--- a/pdp/ast/jast/parser_test.go
+++ b/pdp/ast/jast/parser_test.go
@@ -323,6 +323,32 @@ const (
               }
             ]
           },
+          {
+            "equal": [
+              {
+                "attr": "lsAttr"
+              },
+              {
+                "val": {
+                  "type": "list of strings",
+                  "content": ["first", "second", "third"]
+                }
+              }
+            ]
+          },
+          {
+            "equal": [
+              {
+                "attr": "ssAttr"
+              },
+              {
+                "val": {
+                  "type": "set of strings",
+                  "content": ["first", "second", "third"]
+                }
+              }
+            ]
+          },
           {
             "all": [
               {
```